Feature/wifi

DaSchaef opened the pull request:
Draft pull request for code discussions :)
Not finished yet, untested so far.

DaSchaef commented:
@ravenst0ne @jschroeter 
I think it can be tested now. Currently I have no bus setup here, so my tests were limited so far.

Default WiFi Login:
```
#define DEFAULT_WIFI_SSID     "GDOOR"
#define DEFAULT_WIFI_PASSWORD "12345678"
```

Rest should be self-explaining I hope (captive portal, setup etc)

ravenst0ne commented on firmware/esp32/gdoor/src/mqtt_helper.h:
```diff
@@ -0,0 +1,45 @@
+class MQTT_PRINTER : public Print { // Class/Struct to collect bus related infos
+    public:
+        MQTTClient *mqttClient;
+        char buffer[201];
```
Buffer size too small in case of debug printing

DaSchaef commented:
> One more finding: The HTML title is the same as the header on the page, so right now we have the base64 encoded image twice, but as title it's not possible to decode it. There is already a todo marked in the Wifimanager library to add supoort to change the title separately (https://github.com/tzapu/WiFiManager/blob/e978bc059c522404c01e06cd136fcf23234eb784/WiFiManager.cpp#L1336)
> 
> One workaround would be to add a custom response handler, or better wait until the library
 supports it?

The "WifiManager" Library has even more bugs I worked around, I noticed the title myself, but think it is better to keep as-is and maybe patch it on our own in the future. (it is "only" ugly) - I would not count on them.

jschroeter commented:
Would it be possible to add MQTT username + password? Wanted to try it with Home Assistant but there the default MQTT config requires authentication.

jschroeter commented on firmware/esp32/gdoor/main.cpp:
```diff
@@ -70,25 +75,39 @@ void setup() {
+        if(str_received.length() > 0) {
+            if(!parse(str_received)) { //Check if received string is a command
+                GDOOR::send(str_received); // Send to bus if it is not a command
+                Serial.print("Send: ");
```
These two lines could be removed again, right? Just noticed that it pollutes the serial output.
The other logs like `Serial.println("GDOOR Setup start")` are useful I think and not a big issue as they only appear once when booting.

ravenst0ne replied:
Or wrap it with the debug mode?

jschroeter commented on firmware/esp32/gdoor/src/defines.h:
```diff
@@ -39,4 +39,14 @@
+// MQTT
+#define DEFAULT_MQTT_SERVER   "0.0.0.0" 
+#define DEFAULT_MQTT_PORT     "1883" 
+#define DEFAULT_MQTT_TOPIC_BUS_RX "/gdoor/bus_rx"
```
is the leading slash a common thing? Afaik usually it's without.

jschroeter commented:
Thanks, I can confirm that the MQTT username/password work.

What I initially didn't realise is that I do need to restart the adapter after changing the MQTT config. Could this be updated without a restart? If not, could we add a hint or maybe even a restart button in the UI?

DaSchaef commented:
> Thanks, I can confirm that the MQTT username/password work.
> 
> What I initially didn't realise is that I do need to restart the adapter after changing the MQTT config. Could this be updated without a restart? If not, could we add a hint or maybe even a restart button in the UI?

It should restart:
https://github.com/gdoor-org/gdoor/blob/feature/wifi/firmware/esp32/gdoor/src/wifi_helper.cpp#L319
I need to check tomorrow, if there is a bug.

jschroeter commented:
I'm currently observing something weird which happened twice in the last days:
- GDoor stopped receiving actions, they didn't appear via serial neither MQTT
- Wifi was not connected anymore (router showed the devices as offline & I couldn't connect to the web interface). Maybe related to that we had Wifi off for some hours, but I would expect it just reconnects once the network is available again?
- sending via serial was still working fine

DaSchaef commented:
Very ŝtrange, the mentioned behavior does not align to what I would expect on wifi disconnect.
Can you reproduce it with wifi on/off?

jschroeter commented:
Some more observations since it happened again:

- it seems to be unrelated to the unavailability of wifi, it now also happened while wifi was constantly available
- the adapter successfully reconnected to wifi when I turned wifi off over night as a test
- interestingly different things break
  - as written above, first time I noticed it both serial and MQTT didn't receive events anymore. Also the adapter didn't show connected to the wifi in the router and wasn't accessible through the browser.
  - second time MQTT wasn't receiving events anymore, but the serial interface did
  - today MQTT was still receiving events, but the serial interface didn't
  - in all cases sending actions over serial still worked

Initially I thought it's an issue with the Home Assistant integration, but it seems unlikely that serial and MQTT break at the same time and wouldn't explain why my router showed the adapter as disconnected from the wifi.

DaSchaef commented:
Is the observation via homeassistent or also via e.g. mqttbrowser and a serial console?
We really need to exclude homeassistent as error source. The symptoms are very very strange.

I'm currently preparing a build with more debug output.

DaSchaef commented:
I added more debug outputs (untested, maybe I broke it).

Please log the serial debug prints, it is very verbose,, but maybe we can see what happens.
My only idea would be that the MQTT sending is somehow blocking

Hmmm, if something went wrong with MQTT/WIFI, maybe the MQTT_PRINTER Buffer was too full?
I added big fat debug print statement for such a condition.

jschroeter commented:
Good idea trying it with MQTT Explorer!
I think I found the issue and it's related to Home Assistant. Apparently HA gets confused when sensor values change types, e.g. from valid JSON, to invalid JSON, to string etc. This is true for the serial integration but also for MQTT. And sometimes the sensor doesn't recover, it keeps showing the value "unknown" even if it's a valid JSON again.
For serial it's also important to always terminate with a new line which was missing in two places ([017ccee](https://github.com/gdoor-org/gdoor/pull/10/commits/017ccee442b76e5b6ba1c79e57b96fb76019ab50)).

In general, I think it makes sense to have a consistent output. Let's discuss this in #15. And I would say we can merge this PR.
